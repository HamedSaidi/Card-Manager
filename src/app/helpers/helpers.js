import { isFinite, isArray, isString, groupBy , sumBy, round, toFinite } from 'lodash'

export const toSafeNumber = value => round(toFinite(value), 2)

export const safeSum = (list, key) => {
  if(isArray(list) && isString(key))
    return toSafeNumber(sumBy(list, key))

  return 0
}

export const calculateProductBruttoPrice = ({price = 0, tax = 0} = {price: 0, tax: 0}) => {
  if (isFinite(price) && isFinite(tax))
    return toSafeNumber(price * (1 + 0.01 * tax))

  return 0
}

const calculateTotalBrutto = (products) => sumBy(products, calculateProductBruttoPrice)

export const calculateTotalBruttoByTaxLevels = (products = []) => {
  const productsByTaxLevels = groupBy(products, 'tax')
  const taxationLevels = Object.keys(productsByTaxLevels)

  return taxationLevels.map(tax => ({
    label: tax,
    value: toSafeNumber(calculateTotalBrutto(productsByTaxLevels[tax]))
  }))
}
