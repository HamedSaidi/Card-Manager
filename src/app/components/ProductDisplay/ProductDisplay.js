import React from 'react'
import PropTypes from 'prop-types'

import { calculateProductBruttoPrice } from '../../helpers'

const ProductDisplay = React.memo(({comment, handleDelete, handleEdit, id, name, price, tax}) => (
  <tr key={id}>
    <td className="products-table-element">{id}</td>
    <td className="products-table-element">{name}</td>
    <td className="products-table-element">{comment}</td>
    <td className="products-table-element">{price} &euro;</td>
    <td className="products-table-element">{tax} %</td>
    <td className="products-table-element">{calculateProductBruttoPrice({price, tax})}</td>
    <td className="product-item-btn-container">
      <button onClick={() => handleDelete(id)}>X</button>
      <button onClick={() => handleEdit({id, field: 'editable', value: true})}>Edit</button>
    </td>
  </tr>
))

ProductDisplay.propTypes = {
  comment: PropTypes.string.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleEdit: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  tax: PropTypes.number.isRequired
}

export default ProductDisplay
