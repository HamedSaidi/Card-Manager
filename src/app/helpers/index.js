import { toSafeNumber, safeSum, calculateProductBruttoPrice, calculateTotalBruttoByTaxLevels } from './helpers'

export { toSafeNumber, safeSum, calculateProductBruttoPrice, calculateTotalBruttoByTaxLevels }
