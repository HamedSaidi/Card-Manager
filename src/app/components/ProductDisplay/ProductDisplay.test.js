import React from 'react'
import renderer, {render} from 'react-test-renderer'
import { toString } from 'lodash'
import ProductDisplay from './ProductDisplay'
import { calculateProductBruttoPrice } from '../../helpers'

describe('<ProductDisplay/>', () => {
  it('should render properly', () => {
    const props = {
      id: 1,
      name: 'test name',
      comment: 'test comment',
      price: 21,
      editable: true,
      tax: 0,
      handleEdit: e => true,
      handleDelete: e => true
    }

    const component = renderer.create(<ProductDisplay {...props}  />)

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    expect(tree.children[0].children[0]).toEqual(toString(props.id))
    expect(tree.children[1].children[0]).toEqual(props.name)
    expect(tree.children[2].children[0]).toEqual(props.comment)
    expect(tree.children[3].children[0]).toEqual(toString(props.price))
    expect(tree.children[4].children[0]).toEqual(toString(props.tax))
    expect(tree.children[5].children[0]).toEqual(toString((calculateProductBruttoPrice(props))))
  })

  it('should call handleDelete when user clicks on delete button', () => {
    const handleDelete = jest.fn(e => true)
    const props = {
      id: 1,
      name: 'test name',
      comment: 'test comment',
      price: 21,
      editable: true,
      tax: 0,
      handleEdit: e => true,
      handleDelete
    }

    const component = renderer.create(<ProductDisplay {...props}  />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback for the delete button
    tree.children[6].children[0].props.onClick()
    expect(handleDelete).toHaveBeenCalled()
  })

  it('should call handleEdit when user clicks on delete button', () => {
    const handleEdit = jest.fn(e => true)
    const props = {
      id: 1,
      name: 'test name',
      comment: 'test comment',
      price: 21,
      editable: true,
      tax: 0,
      handleEdit,
      handleDelete: e => true
    }

    const component = renderer.create(<ProductDisplay {...props}  />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback for the delete button
    tree.children[6].children[1].props.onClick()
    expect(handleEdit).toHaveBeenCalled()
  })
})
