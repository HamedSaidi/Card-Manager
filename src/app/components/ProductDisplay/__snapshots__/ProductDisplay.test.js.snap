// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDisplay/> should call handleDelete when user clicks on delete button 1`] = `
<tr>
  <td
    className="products-table-element"
  >
    1
  </td>
  <td
    className="products-table-element"
  >
    test name
  </td>
  <td
    className="products-table-element"
  >
    test comment
  </td>
  <td
    className="products-table-element"
  >
    21
     €
  </td>
  <td
    className="products-table-element"
  >
    0
     %
  </td>
  <td
    className="products-table-element"
  >
    21
  </td>
  <td
    className="product-item-btn-container"
  >
    <button
      onClick={[Function]}
    >
      X
    </button>
    <button
      onClick={[Function]}
    >
      Edit
    </button>
  </td>
</tr>
`;

exports[`<ProductDisplay/> should call handleEdit when user clicks on delete button 1`] = `
<tr>
  <td
    className="products-table-element"
  >
    1
  </td>
  <td
    className="products-table-element"
  >
    test name
  </td>
  <td
    className="products-table-element"
  >
    test comment
  </td>
  <td
    className="products-table-element"
  >
    21
     €
  </td>
  <td
    className="products-table-element"
  >
    0
     %
  </td>
  <td
    className="products-table-element"
  >
    21
  </td>
  <td
    className="product-item-btn-container"
  >
    <button
      onClick={[Function]}
    >
      X
    </button>
    <button
      onClick={[Function]}
    >
      Edit
    </button>
  </td>
</tr>
`;

exports[`<ProductDisplay/> should render properly 1`] = `
<tr>
  <td
    className="products-table-element"
  >
    1
  </td>
  <td
    className="products-table-element"
  >
    test name
  </td>
  <td
    className="products-table-element"
  >
    test comment
  </td>
  <td
    className="products-table-element"
  >
    21
     €
  </td>
  <td
    className="products-table-element"
  >
    0
     %
  </td>
  <td
    className="products-table-element"
  >
    21
  </td>
  <td
    className="product-item-btn-container"
  >
    <button
      onClick={[Function]}
    >
      X
    </button>
    <button
      onClick={[Function]}
    >
      Edit
    </button>
  </td>
</tr>
`;
